### ISSUE TITLE
Non-deterministic results across machines despite fixed seeds (fine-tuning)

### ISSUE THREAD

**Issue by erfannsr** [v-iashin/Synchformer#16]
When fine-tuning the model on LRS2 for lipsync evaluation, I noticed that training on different machines (with the same dataset, architecture, and fixed random seed) produces different results during testing.

  - Same dataset, preprocessing, architecture, and hyperparameters.
  - Seeds fixed across `random`, `numpy`, and `torch` (CPU + CUDA).

- **Machines:**
  - Machine A: NVIDIA RTX A4000, CUDA 12.4
  - Machine B: NVIDIA GeForce RTX 3090, CUDA 11.7

- **Dependencies (common on both):**
  - torch 2.0.1  
  - torchaudio 2.0.2  
  - torchvision 0.15.2  
  - ffmpeg-python 0.2.0  
  - av 10.0.0  

- **Observation:**
 Training diverges across the two machines and produces different test results, even with the same random seed and identical environments.
  This suggests non-determinism in training the model.

- **Question:**
1. Is this expected behavior in your implementation, or should results be reproducible across machines?
2. Are there known limitations or recommended practices in this repo for ensuring cross-machine determinism?

Thanks for your great work in this repo, and I’d appreciate any guidance to make my training reproducible across setups.

**Comment by v-iashin** [comment]
1. I didn't test training across different machines. So, I'm not sure. Could you share your training curves (train loss, eval accuracy) from a) two machines; b) two runs on one machine? 
2. Since I didn't notice it before, the repo does not include any. If you take one of your (or my) checkpoints, does both machines output exactly the same results for a input?

**Comment by erfannsr** [comment]
1. Please find the training curves below:
these are from training on two machines (a). I cannot yet provide the second case, as the second run on the machine is still ongoing (recently started).
<img width="343" height="235" alt="Image" src="https://github.com/user-attachments/assets/67de3cff-4751-4f3b-8b52-cb2768d758d7" />
<img width="347.5" height="235" alt="Image" src="https://github.com/user-attachments/assets/0421e0a8-069b-46ed-805a-b1b7b2647fa7" />
<img width="350" height="235" alt="Image" src="https://github.com/user-attachments/assets/a95478ab-922e-421b-b34d-4b58e05f82f6" />


2. Yes, I checked and verified both machines produce the same output using a common checkpoint.

**Comment by v-iashin** [comment]
Something is seriously off. You are getting 85% training accuracy at 20 epochs on B vs 65% on A. It is unlikely that a machine setup alone could cause this. 

The shape of the plots suggest that the machine B trains the model 2x more efficiently than A (80% train accuracy at 30 epochs vs at 15 epochs). Sometimes 2x batch size can cause that or a 2x smaller dataset.

I am leaning towards suggesting that you may have a bug somewhere in your setup: a) in your data; b) code is not exactly the same on both machines; c) weight initialisation is not working correctly on A (check for places with `strict=False` and ensure checkpoints on A and B have the same `md5sum`).

**Comment by erfannsr** [comment]
Oh, I need to make a clarification on this.

The training on machine B is actually running on 3 GPUs in parallel. I believe that's why it's training more efficiently. Can that cause any significant mismatch in the final training results?

I checked the weight initializations on both machines and did not observe any difference in initial weights.

For context, I’ve added a cross-attention module on top of the aud/vid feature extractors. Because of that, I had to use `strict=False` when loading the checkpoint. Initially, I thought this might cause the additional layers to be initialized differently across machines, but since the random seed is fixed, the initialized weights turn out to be consistent (I verified they match).

Do you think there could be another factor I might be overlooking here? I’d greatly appreciate any further thoughts or suggestions.
